dht: convert ring_position and decorated_key to std::strong_ordering

avikivity opened the pull request:
As #1449 notes, trichotomic comparators returning int are dangerous as they
can be mistaken for less comparators. This series converts dht::ring_position
and dht::decorated_key, as well as a few closely related downstream types, to
return std::strong_ordering.

avikivity commented:
Still in draft until it can be rebased on top of #8222.

avikivity commented:
Rebased on top of #8222, no other changes.

nyh commented on utils/collection-concepts.hh:
```diff
@@ -41,13 +42,18 @@ SEASTAR_CONCEPT(
+// Temporary scaffolding for supporting trichotomic compares
+// that return either int or std::strong_ordering
+template <typename T>
+concept int_or_strong_ordering = std::same_as<T, int> || std::same_as<T, std::strong_ordering>;
```
I'm a little unsure on what our coding style is - what's the point of using the "concept" keyword here without any old-compiler protection, and still use the SEASTAR_CONCEPT macro two lines below?
If the decision is that we don't need SEASTAR_CONCEPT any more, maybe you should remove it below? I know it's preexisting, but since you're adding a new concept anyway, it's a very good time.

avikivity replied:
Scylla requires C++20, which comes with concepts.

I'm happy to follow up with a patch that removes SEASTAR_CONCEPT globally, but would rather not touch unrelated code in this patch.

nyh replied:
Well, I obviously don't expect you to remove SEASTAR_CONCEPT globally, but here you are specifically creating a new concept to be used one line down, and yet the one line down is "protected" by SEASTAR_CONCEPT, but the new concept isn't. I think it's a good opportunity to at least clean this line. The mixture in the current patch is just strange.

avikivity replied:
I don't see how the patch is mixing anything. It's certainly not our style to have drive-by fixes.

nyh replied:
Well, unless I'm misreading it, your patch takes code which supports old compilers (because it's "protected" by SEASTAR_CONCEPT) and then actively breaks this support by adding a new concept outside the SEASTAR_CONCEPT. I think you need to decide whether you want to keep this support (and hide the new concept as well) or if you don't want to keep it, but the combination doesn't make sense.

avikivity replied:
https://github.com/scylladb/scylla/pull/8236

nyh commented on test/boost/mutation_test.cc:
```diff
@@ -2586,6 +2586,23 @@ void merge_container(
     }
 }
 
+template <typename Container, typename OutputIt>
```
I would suggest to leave a comment before this version saying it is left for old code, and that new code should use the strong_ordering variant above.

avikivity replied:
Ok, will add.

nyh commented on test/boost/total_order_check.hh:
```diff
@@ -22,23 +22,27 @@
 private:
-    static int sgn(int x) {
+    static int sgn(std::strong_ordering x) {
         return (x > 0) - (x < 0);
```
I guess it's ok (it's preexisting code which you just changed the parameter type), but wouldn't "x <=> 0" work instead of this wierd calculation?

avikivity replied:
No. x <=> returns an std::strong_ordering value.

nyh commented on test/boost/total_order_check.hh:
```diff
@@ -22,23 +22,27 @@
     }
-
+    static int sgn(int x) {
+        return sgn(x <=> 0);
```
Wouldn't just "return x <=> 0" work? Why do you need to call sgn() on it? Maybe I'm missing something :-(

avikivity replied:
Yes. The return value is int, not std::strong_ordering.

nyh commented on test/boost/total_order_check.hh:
```diff
@@ -22,23 +22,27 @@
+    }
     // All in left are expect to compare with all in right with expected_order result
-    void check_order(const std::vector<element>& left, const std::vector<element>& right, int order) {
+    void check_order(const std::vector<element>& left, const std::vector<element>& right, int_or_strong_ordering auto order) {
```
Maybe this is a stupid question (I'm not fluent in concepts yet...), is this "auto" necessary here? Doesn't "int_or_strong_ordering order" work without the "auto"?

denesb replied:
`int_or_strong_ordering` is a concept, which works on templates. This function is now actually a template (because of the `auto`) and the aforementioned concept restricts the accepted types.

denesb replied:
A more verbose way of achieving the same would be:
```c++
template <typename Order>
requires int_or_strong_ordering<Order>
void check_order(const std::vector<element>& left, const std::vector<element>& right, Order order) {
```

nyh commented on dht/i_partitioner.cc:
```diff
@@ -295,43 +295,43 @@ bool ring_position::less_compare(const schema& s, const ring_position& other) co
     }
     if (lh._key && rh._key) {
         auto c = lh._key->legacy_tri_compare(s, *rh._key);
         if (c) {
```
In a previous patch you fixed if's like if(!c) to be if(c==0) to make it clear this is not a boolean but a strong ordering. Isn't it prudent to do the same with "if(c)" here, making it "if(c != 0)" to make it clearer this is not a boolean?

avikivity replied:
This will get fixed when I patch legacy_tri_compare() to return std::strong_ordering. I'd rather not convert the entire code base at the same time.

nyh commented on dht/i_partitioner.cc:
```diff
@@ -295,43 +295,43 @@ bool ring_position::less_compare(const schema& s, const ring_position& other) co
         auto c = lh._key->legacy_tri_compare(s, *rh._key);
         if (c) {
-            return c;
+            return c <=> 0;
```
Wouldn't "return c" do exactly the same? Or did you deliberately not convert legacy_tri_convert() and c is an integer?

avikivity replied:
Yes, it is an integer.

nyh commented on dht/i_partitioner.cc:
```diff
@@ -295,43 +295,43 @@ bool ring_position::less_compare(const schema& s, const ring_position& other) co
         auto rel = rh.key().tri_compare(s, *lh._key);
         if (rel) {
-            return -rel;
+            return -rel < 0 ? std::strong_ordering::less : std::strong_ordering::greater;
```
I was curious if there's a less ugly way to do this, and I think there is:
return 0 <=> rel;
Note the zero is first, so the directions are reversed.
EDIT: I see you used exactly this trick a few lines down.

avikivity replied:
Yes, I'll change this too.

avikivity replied:
Changed.

nyh commented on repair/repair.hh:
```diff
@@ -357,7 +357,11 @@ struct repair_sync_boundary {
+            auto ret = 0;
+            if (tmp != 0) {
+                ret = tmp < 0 ? -1 : 1;
+            }
```
Instead of the four lines above, wouldn't just this work?
auto ret = tmp <=> 0

If it works, you don't even need to name tmp, just do

auto ret = _pk_cmp(a.pk, b.pk) <=> 0'

What am I missing?

avikivity replied:
The return value of <=> cannot be assigned to an int.

nyh reviewed (changes requested):
Looks good. I had a few comments, questions and suggestions, but none of them is critical (I didn't find any error in what you did) and probably some of them are just my misunderstanding :-) But still, maybe you'll still want to consider some of them.

avikivity commented:
Updated:
 - use 0 <=> rel instead of conditional operator (?:)
 - add comment to compatibility variant of merge_container()

avikivity commented:
@nyh or @tgrabiec, ping

psarna commented:
Dequeuing due to the fact that after another dequeue (`reader_concurrency_semaphore: limit non-admitted inactive reads`), this series no longer compiles on top of next.

avikivity commented:
> Dequeuing due to the fact that after another dequeue (`reader_concurrency_semaphore: limit non-admitted inactive reads`), this series no longer compiles on top of next.

It was me re-merging it incorrectly while dropping something else. I'll rebase.

avikivity commented:
Rebased to resolve conflict with 9038a81317dffb1b4d7cbbc5d45fa25d16609d97. The original merge resolved it correctly, but when I dropped another bad series, I didn't re-merge correctly.

psarna commented:
requeued
